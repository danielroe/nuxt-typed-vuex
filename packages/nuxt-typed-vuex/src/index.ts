import { join, resolve } from 'upath'

import type { Module, NuxtOptions } from '@nuxt/types'

import { name, version } from '../package.json'

/**
 * @private
 */
const nuxtTypedVuex: Module = async function() {
  const nuxtOptions = this.nuxt.options as NuxtOptions

  if (!nuxtOptions.store) console.warn('You do not have a store defined.')

  this.addPlugin({
    src: resolve(__dirname, '../template/plugin.js'),
    fileName: 'nuxt-typed-vuex.js',
    options: {
      store: join(nuxtOptions.buildDir, 'store'),
    },
  })

  nuxtOptions.build.transpile = /* istanbul ignore next */ nuxtOptions.build.transpile || []
  nuxtOptions.build.transpile.push(/typed-vuex/)
}

;(nuxtTypedVuex as any).meta = { name, version }

export default nuxtTypedVuex
