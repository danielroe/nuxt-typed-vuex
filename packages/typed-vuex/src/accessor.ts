import { Store, GetterTree, MutationTree, ActionTree } from 'vuex'
import { NuxtStoreInput, MergedStoreType, BlankStore } from './types/store'
import { State, StateType } from './types/state'
import { NuxtModules } from './types/modules'

export const getAccessorType = <
  T extends State,
  G extends GetterTree<StateType<T>, any>,
  M extends MutationTree<StateType<T>>,
  A extends ActionTree<StateType<T>, any>,
  S extends NuxtModules
>(
  store: Partial<NuxtStoreInput<T, G, M, A, S>>
) => {
  return {} as MergedStoreType<typeof store & BlankStore>
}

const getNestedState = (parent: any, namespaces: string[]): any => {
  if (!parent[namespaces[0]]) {
    return parent
  } else {
    return getNestedState(parent[namespaces[0]], namespaces.slice(1))
  }
}

const createAccessor = <T extends State, G, M, A, S extends NuxtModules>(
  store: Store<any>,
  {
    getters,
    state,
    mutations,
    actions,
    namespaced,
  }: Partial<NuxtStoreInput<T, G, M, A, S>>,
  namespace = ''
) => {
  const namespacedPath = namespace && namespaced ? `${namespace}/` : ''
  const accessor: Record<string, any> = {}
  Object.keys(getters || {}).forEach(getter => {
    Object.defineProperty(accessor, getter, {
      get: () => store.getters[`${namespacedPath}${getter}`],
    })
  })
  const evaluatedState = state
    ? typeof state === 'function'
      ? state()
      : state
    : {}
  Object.keys(evaluatedState).forEach(prop => {
    if (!Object.getOwnPropertyNames(accessor).includes(prop)) {
      const namespaces = namespace.split('/')
      Object.defineProperty(accessor, prop, {
        get: () => getNestedState(store.state, namespaces)[prop],
      })
    }
  })
  Object.keys(mutations || {}).forEach(mutation => {
    accessor[mutation] = (mutationPayload: any) =>
      store.commit(`${namespacedPath}${mutation}`, mutationPayload)
  })
  Object.keys(actions || {}).forEach(action => {
    accessor[action] = (actionPayload: any) =>
      store.dispatch(`${namespacedPath}${action}`, actionPayload)
  })
  return accessor
}

export const useAccessor = <
  T extends State,
  G extends GetterTree<StateType<T>, any>,
  M extends MutationTree<StateType<T>>,
  A extends ActionTree<StateType<T>, any>,
  S extends NuxtModules
>(
  store: Store<any>,
  input: Partial<NuxtStoreInput<T, G, M, A, S>>,
  namespace?: string
) => {
  const accessor = createAccessor(store, input, namespace)
  Object.keys(input.modules || {}).forEach(moduleNamespace => {
    const nestedNamespace = namespace
      ? `${namespace}/${moduleNamespace}`
      : moduleNamespace
    accessor[moduleNamespace] = useAccessor(
      store,
      (input.modules as any)[moduleNamespace],
      nestedNamespace
    )
  })

  const storeType = getAccessorType(input)

  return accessor as typeof storeType
}

export const getAccessorFromStore = (pattern: any) => {
  return (store: Store<any>) =>
    useAccessor(store, pattern._modules.root._rawModule)
}
