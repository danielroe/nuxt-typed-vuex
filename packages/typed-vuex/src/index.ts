export { getStoreType } from './types/store'
export { actionTree, NormalisedActionHandler } from './types/actions'
export { getterTree } from './types/getters'
export { mutationTree, Commit } from './types/mutations'
export { useAccessor, getAccessorType, getAccessorFromStore } from './accessor'
