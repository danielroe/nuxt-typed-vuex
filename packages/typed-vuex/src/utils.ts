/* eslint-disable @typescript-eslint/ban-ts-comment */
import { Mapper } from './types/utils'

export const createMapper = <T extends Record<string, any>>(accessor: T) => {
  const mapper: Mapper<T> = (prop: any, properties?: string[]) => {
    if (!properties) {
      return Object.fromEntries(
        (Array.isArray(prop) ? prop : [prop]).map(property => [
          property,
          function(...args: any[]) {
            const value = accessor
              ? accessor[property]
              : // @ts-ignore
                this.$accessor[property]
            if (value && typeof value === 'function') return value(...args)
            return value
          },
        ])
      )
    }
    return Object.fromEntries(
      properties.map(property => [
        property,
        function(...args: any[]) {
          const value =
            // @ts-ignore
            accessor ? accessor[prop][property] : this.$accessor[prop][property]
          if (value && typeof value === 'function') return value(...args)
          return value
        },
      ])
    )
  }
  return mapper
}
