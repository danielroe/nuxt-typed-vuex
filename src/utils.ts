import { VueConstructor } from 'vue/types/vue'
import Vuex, {
  Store,
  GetterTree,
  MutationTree,
  DispatchOptions,
  CommitOptions,
  StoreOptions,
} from 'vuex'

declare module 'vue' {
  interface VueConstructor {
    $accessor?: any;
  }
}

export class TypedStore<S> extends Store<S> {
  $accessor: VueConstructor['$accessor']
}
type UnifiedStore<S> = TypedStore<S> | Store<S>
export type TypedStoreOptions<S> = Omit<StoreOptions<S>, 'actions'> & {
  actions?: ModifiedActionTree<any>;
}

type Not<T, M> = T extends M ? never : T

type StateObject = Not<Record<string, any>, Function>
type StateFunction = Not<() => unknown | any, Record<string, any>>
type State = StateObject | StateFunction
type StateType<T extends State> = T extends () => any ? ReturnType<T> : T

type MergedFunctionProcessor<T extends () => any, O> = Parameters<
  T
>[1] extends undefined
  ? (options?: O) => ReturnType<T>
  : (payload: Parameters<T>[1], options?: O) => ReturnType<T>

type GettersTransformer<T extends Record<string, any>> = Readonly<
  { [P in keyof T]: ReturnType<T[P]> }
>
type MutationsTransformer<T extends Record<string, any>> = {
  [P in keyof T]: MergedFunctionProcessor<T[P], CommitOptions>
}
type ActionTransformer<T extends Record<string, any>> = {
  [P in keyof T]: MergedFunctionProcessor<T[P], DispatchOptions>
}
type ModuleTransformer<T, O = string> = T extends NuxtModules
  ? { [P in keyof T]: MergedStoreType<T[P] & BlankStore, O> }
  : {}

interface BlankStore {
  state: {};
  getters: {};
  mutations: {};
  actions: {};
  modules: {};
}

interface NuxtStore {
  state: State;
  getters: Record<string, any>;
  mutations: Record<string, any>;
  actions: Record<string, any>;
  modules: NuxtModules;
}
type NuxtModules = Record<string, Partial<NuxtStore>>

interface NuxtStoreInput<
  T extends State,
  G,
  M,
  A,
  S extends { [key: string]: Partial<NuxtStore> }
> {
  state: T;
  getters?: G;
  mutations?: M;
  actions?: A;
  modules?: S;
}

type MergedStoreType<T extends NuxtStore, K = string> = ('state' extends K
  ? StateType<T['state']>
  : {}) &
  ('getters' extends K ? GettersTransformer<T['getters']> : {}) &
  ('mutations' extends K ? MutationsTransformer<T['mutations']> : {}) &
  ('actions' extends K ? ActionTransformer<T['actions']> : {}) &
  ('modules' extends K ? ModuleTransformer<T['modules']> : {})

type StoreParameter<T extends () => any> = Parameters<T>[1] extends undefined
  ? never
  : Parameters<T>[1]

interface Dispatch<T extends Record<string, () => any>> {
  <P extends keyof T>(
    action: P,
    payload: StoreParameter<T[P]>,
    options?: DispatchOptions
  ): ReturnType<T[P]>;
  <P extends keyof T>(
    action: StoreParameter<T[P]> extends never ? P : never,
    options?: DispatchOptions
  ): ReturnType<T[P]>;
}

interface Commit<T extends Record<string, () => any>> {
  <P extends keyof T>(
    mutation: P,
    payload: StoreParameter<T[P]>,
    options?: DispatchOptions
  ): ReturnType<T[P]>;
  <P extends keyof T>(
    mutation: StoreParameter<T[P]> extends never ? P : never,
    options?: CommitOptions
  ): ReturnType<T[P]>;
}

export type ActionContext<T extends Required<NuxtStore>> = {
  state: StateType<T['state']>;
  getters: { [P in keyof T['getters']]: ReturnType<T['getters'][P]> };
  commit: Commit<T['mutations']>;
  dispatch: Dispatch<T['actions']>;
  rootState: any;
  rootGetters: any;
}

export type RootStateHelper<T extends Required<NuxtStore>> = StateType<
  T['state']
> &
  ModuleTransformer<T['modules'], 'state'>

export type RootGettersHelper<
  T extends Required<NuxtStore>
> = GettersTransformer<T['getters']> &
  ModuleTransformer<T['modules'], 'getters'>

export const getStoreType = <T extends State, G, M, A, S extends NuxtModules>(
  store: NuxtStoreInput<T, G, M, A, S>
) => {
  return {
    actionContext: {} as ActionContext<typeof store & BlankStore>,
    rootState: {} as RootStateHelper<typeof store & BlankStore>,
    rootGetters: {} as RootGettersHelper<typeof store & BlankStore>,
    storeInstance: {} as ActionContext<typeof store & BlankStore> &
      Exclude<Store<StateType<T>>, ActionContext<any>>,
  }
}

export const getAccessorType = <
  T extends State,
  G extends GetterTree<StateType<T>, any>,
  M extends MutationTree<StateType<T>>,
  A extends ModifiedActionTree<any>,
  S extends NuxtModules
>(
  store: Partial<NuxtStoreInput<T, G, M, A, S>>
) => {
  return {} as MergedStoreType<typeof store & BlankStore>
}

const getNestedState = (parent: any, namespaces: string[]): any => {
  if (!parent[namespaces[0]]) {
    return parent
  } else {
    return getNestedState(parent[namespaces[0]], namespaces.slice(1))
  }
}

const createAccessor = <T extends State, G, M, A, S extends NuxtModules>(
  store: UnifiedStore<any>,
  {
    getters,
    state,
    mutations,
    actions,
  }: Partial<NuxtStoreInput<T, G, M, A, S>>,
  namespace = ''
) => {
  const namespacedPath = namespace ? `${namespace}/` : ''
  const accessor: Record<string, any> = {}
  Object.keys(getters || {}).forEach(getter => {
    Object.defineProperty(accessor, getter, {
      get: () => store.getters[`${namespacedPath}${getter}`],
    })
  })
  Object.keys(
    state ? (typeof state === 'function' ? state() : state) : {}
  ).forEach(prop => {
    if (!Object.getOwnPropertyNames(accessor).includes(prop)) {
      const namespaces = namespacedPath.split('/')
      const state = getNestedState(store.state, namespaces)
      Object.defineProperty(accessor, prop, {
        get: () => state[prop],
      })
    }
  })
  Object.keys(mutations || {}).forEach(mutation => {
    accessor[mutation] = (mutationPayload: any) =>
      store.commit(`${namespacedPath}${mutation}`, mutationPayload)
  })
  Object.keys(actions || {}).forEach(action => {
    accessor[action] = (actionPayload: any) =>
      store.dispatch(`${namespacedPath}${action}`, actionPayload)
  })
  return accessor
}

export const useAccessor = <
  T extends State,
  G extends GetterTree<StateType<T>, any>,
  M extends MutationTree<StateType<T>>,
  A extends ModifiedActionTree<any>,
  S extends NuxtModules
>(
  store: UnifiedStore<any>,
  input: Partial<NuxtStoreInput<T, G, M, A, S>>,
  namespace?: string
) => {
  const accessor = createAccessor(store, input, namespace)
  Object.keys(input.modules || {}).forEach(moduleNamespace => {
    const nestedNamespace = namespace
      ? `${namespace}/${moduleNamespace}`
      : moduleNamespace
    accessor[moduleNamespace] = useAccessor(
      store,
      (input.modules as any)[moduleNamespace],
      nestedNamespace
    )
  })

  const storeType = getAccessorType(input)

  return accessor as typeof storeType
}

export const getAccessorFromStore = (pattern: any) => {
  return (store: Store<any>) =>
    useAccessor(store, pattern._modules.root._rawModule)
}

export const getterTree = <S, T extends GetterTree<StateType<S>, any>>(
  _state: S,
  tree: T
) => tree

export const mutationTree = <S, T extends MutationTree<StateType<S>>>(
  _state: S,
  tree: T
) => tree

interface ActionHandler<T extends NuxtStore> {
  (
    this: TypedStore<StateType<T['state']>>,
    injectee: Omit<ActionContext<T>, 'dispatch'>,
    payload?: any
  ): any;
}

interface ModifiedActionTree<T extends NuxtStore> {
  [key: string]: ActionHandler<T>;
}

export const actionTree = <
  S extends State,
  G extends GetterTree<StateType<S>, any>,
  M extends MutationTree<StateType<S>>,
  T extends ModifiedActionTree<Required<NuxtStoreInput<S, G, M, {}, {}>>>
>(
  _store: NuxtStoreInput<S, G, M, {}, {}>,
  tree: T
) => tree

export const createStore = <S>(
  VuexModule: typeof Vuex,
  options: TypedStoreOptions<S>
) => {
  return new VuexModule.Store((options as unknown) as StoreOptions<S>)
}

export const registerStoreAccessor = (VuexModule: typeof Vuex, accessor: any) =>
  ((VuexModule.Store as typeof TypedStore).prototype.$accessor = accessor)
