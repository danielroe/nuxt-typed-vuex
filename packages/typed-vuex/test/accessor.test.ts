import { useAccessor, getAccessorType, getAccessorFromStore } from '../src'
import Vuex, { Store } from 'vuex'
import Vue from 'vue'

import { getters, state, actions, mutations } from './fixture'
import * as submodule from './fixture/submodule'

const pattern = {
  getters,
  state,
  actions,
  mutations,
  modules: {
    submodule: {
      ...submodule,
      namespaced: true,
      modules: {
        nestedSubmodule: {
          ...submodule,
          namespaced: true,
        },
      },
    },
  },
}

const accessorType = getAccessorType(pattern)
const submoduleAccessorType = getAccessorType(submodule)

const submoduleBehaviour = (accessor: typeof submoduleAccessorType) => {
  expect(accessor.firstName).toEqual('')
  accessor.setFirstName('Nina')
  expect(accessor.firstName).toEqual('Nina')
  accessor.setLastName('Willis')
  expect(accessor.fullName).toEqual('Nina Willis')

  accessor.initialise()
  expect(accessor.fullName).toEqual('John Baker')
  accessor.setName('Jordan Lawrence')
  expect(accessor.lastName).toEqual('Lawrence')
  accessor.setName('George')
  expect(accessor.firstName).toEqual('George')
}

describe('accessor', () => {
  let store: Store<any>
  let accessor: typeof accessorType

  beforeEach(() => {
    Vue.use(Vuex)
    store = new Store(pattern)
    accessor = useAccessor(store, pattern)
  })
  test('store exists', async () => {
    expect(pattern).toMatchSnapshot()
  })
  test('accessor object exists', async () => {
    expect(accessor).toMatchSnapshot()
  })
  test('accessor can be generated from store object', async () => {
    const accessorFromStore = getAccessorFromStore(store)(store)
    expect(accessorFromStore).toMatchSnapshot()
  })
  test('accessor works with state object', async () => {
    const stateAccessor = useAccessor(store, {
      state: {
        email: 'this does not matter',
      },
      mutations: {
        setEmail(state, newValue: string) {
          ;(state as any).email = newValue
        },
      },
    })
    expect(stateAccessor.email).toMatchSnapshot()
    stateAccessor.setEmail('test1@email.com')
    expect(stateAccessor.email).toMatchSnapshot()
  })
  test('accessor state, getter, mutation and actions work', async () => {
    expect(accessor.fullEmail).toEqual('')
    expect(accessor.initialiseStore()).toBeUndefined()
    accessor.resetEmailWithOptionalPayload('j@j.com')
    expect(accessor.fullEmail).toEqual('j@j.com')
    accessor.resetEmailWithOptionalPayload()
    expect(accessor.fullEmail).toEqual('a@a.com')

    accessor.setEmail('john@j.com')
    expect(accessor.email).toEqual('john@j.com')
    expect(accessor.fullEmail).toEqual('john@j.com')

    accessor.resetEmail()
    expect(accessor.fullEmail).toEqual('a@a.com')
  })
  test('accessor submodule state, getter, mutation and actions work', async () => {
    submoduleBehaviour(accessor.submodule)
  })
  test('namespaced modules work', async () => {
    const nonNamespacedPattern = {
      getters,
      state,
      actions,
      mutations,
      modules: {
        submodule: {
          ...submodule,
          namespaced: false,
        },
      },
    }
    const nonNamespacedStore = new Store(nonNamespacedPattern)
    const nonNamespacedAccessor = useAccessor(
      nonNamespacedStore,
      nonNamespacedPattern
    )
    submoduleBehaviour(nonNamespacedAccessor.submodule)
  })
  test('nested modules work', async () => {
    submoduleBehaviour(accessor.submodule.nestedSubmodule)
  })
  test('namespaced dynamic modules work', async () => {
    store.registerModule('dynamicSubmodule', { ...submodule, namespaced: true })
    const dynamicAccessor = useAccessor(store, {
      modules: { dynamicSubmodule: { ...submodule, namespaced: true } },
    })

    submoduleBehaviour(dynamicAccessor.dynamicSubmodule)

    store.unregisterModule('dynamicSubmodule')
  })
  test('dynamic module helper function works', async () => {
    store.registerModule('dynamicSubmodule', submodule)
    const dynamicAccessor = useAccessor(store, submodule, 'dynamicSubmodule')

    submoduleBehaviour(dynamicAccessor)
    store.unregisterModule('dynamicSubmodule')
  })
})
