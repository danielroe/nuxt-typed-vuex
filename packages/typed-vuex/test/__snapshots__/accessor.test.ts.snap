// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accessor accessor can be generated from store object 1`] = `
{
  "initialiseStore": [Function],
  "resetEmail": [Function],
  "resetEmailWithOptionalPayload": [Function],
  "setEmail": [Function],
  "submodule": {
    "initialise": [Function],
    "nestedSubmodule": {
      "initialise": [Function],
      "setFirstName": [Function],
      "setLastName": [Function],
      "setName": [Function],
    },
    "setFirstName": [Function],
    "setLastName": [Function],
    "setName": [Function],
  },
}
`;

exports[`accessor accessor object exists 1`] = `
{
  "initialiseStore": [Function],
  "resetEmail": [Function],
  "resetEmailWithOptionalPayload": [Function],
  "setEmail": [Function],
  "submodule": {
    "initialise": [Function],
    "nestedSubmodule": {
      "initialise": [Function],
      "setFirstName": [Function],
      "setLastName": [Function],
      "setName": [Function],
    },
    "setFirstName": [Function],
    "setLastName": [Function],
    "setName": [Function],
  },
}
`;

exports[`accessor accessor works with state object 1`] = `""`;

exports[`accessor accessor works with state object 2`] = `"test1@email.com"`;

exports[`accessor store exists 1`] = `
{
  "actions": {
    "resetEmail": [Function],
    "resetEmailWithOptionalPayload": [Function],
  },
  "getters": {
    "email": [Function],
    "fullEmail": [Function],
  },
  "modules": {
    "submodule": {
      "actions": {
        "initialise": [Function],
        "setName": [Function],
      },
      "getters": {
        "fullName": [Function],
      },
      "modules": {
        "nestedSubmodule": {
          "actions": {
            "initialise": [Function],
            "setName": [Function],
          },
          "getters": {
            "fullName": [Function],
          },
          "mutations": {
            "setFirstName": [Function],
            "setLastName": [Function],
          },
          "namespaced": true,
          "state": [Function],
        },
      },
      "mutations": {
        "setFirstName": [Function],
        "setLastName": [Function],
      },
      "namespaced": true,
      "state": [Function],
    },
  },
  "mutations": {
    "initialiseStore": [Function],
    "setEmail": [Function],
  },
  "state": [Function],
}
`;
